### PR TITLE
Support configuration services: inline objects

### PR DESCRIPTION
**Pull request by uncleDecart**
This is PR has two main parts:

- It refactors zedrouter's metadata server and switch from http mux to go-chi router
- It adds first route to send `description.json` file to metadata server
- It adds parsing  and publishing of `PatchEnvelope` structure to zedagent

It's still work in progress, but I'd be glad to get some feedback :) 

It has not yet been tested with controller, will provide updates here

**Edit**:
- `persistcache` library is now part of EVE
- https://github.com/lf-edge/eve-api/pull/12 is merged
- Downloading route is added

### CODE REVIEW

File: pkg/pillar/cmd/zedrouter/metadata.go [hunk 1]
```diff
@@ -8,84 +8,13 @@
...
-	"encoding/json"
-	"fmt"
-	"io"
-	"net"
 	"net/http"
-	"os"
-	"path"
-	"strings"
 
+	"github.com/go-chi/chi/v5"
```
[the remark this answers is not in the document]
Comment (uncleDecart) [reply pkg/pillar/cmd/zedrouter/metadata.go:13]: at some point gorilla was deprecated and had no active maintaners, I heard they found somebody, but better be safe than sorry, plus switching to go-chi costs us nothing since it is based on http.Mux and since we don't have tests for older routes, but it would be nice to have some kind of modern router (because we don't want to parse URL to get route parameters such as /api/v1/{patch_envelope_id}/{file_path}

[the remark this answers is not in the document]
Comment (christoph-zededa) [reply pkg/pillar/cmd/zedrouter/metadata.go:13]: Okay, fine, can you then please remove `gorilla/mux` from `go.{mod,sum}`?

[the remark this answers is not in the document]
Comment (uncleDecart) [reply pkg/pillar/cmd/zedrouter/metadata.go:13]: We do use it [wstunnelclient](https://github.com/lf-edge/eve/blob/master/pkg/pillar/zedcloud/wstunnelclient.go) so it might be a good conversation to think should we switch to gorilla and how we should switch gorilla websockets

[the remark this answers is not in the document]
Comment (christoph-zededa) [reply pkg/pillar/cmd/zedrouter/metadata.go:13]: IMO we should not use both, because then we basically double the risk that one project gets deprecated and we have to fix it and you basically said "better be safe than sorry".

(Ideally we abstract it away from the underlying muxer so we have to change it only at one place, but that's too much for now, I think)

[the remark this answers is not in the document]
Comment (eriknordmark) [reply pkg/pillar/cmd/zedrouter/metadata.go:13]: @uncleDecart Do we need anything fancy from the router necessating switching from the basic and standard http.NewServeMux?

File: pkg/pillar/pubsub/util.go [hunk 2]
```diff
@@ -15,6 +15,15 @@ import (
 )
 
 // deepCopy returns the same type as what is passed as input
+// Warning: only public fields will be exported
+// Note why json marshalling is used:
+// Type casting and associated type assertions in golang are only
+// useful for atoms in the type system. Hence you can't do a type
+// assertion and cast fo a struct of internal fields. This coupled
```
Comment (christoph-zededa) [inline pkg/pillar/pubsub/util.go:22]:
```suggestion
// assertion and cast of a struct of internal fields. This coupled
```

File: pkg/pillar/cmd/zedrouter/metadata.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (uncleDecart) [reply pkg/pillar/cmd/zedrouter/metadata.go:13]: Reason why I'm bringing go-chi here is because I want to use path parameters which, i.e. 
`/eve/v1/patch/{patch-id}/{file-name}`
It is not yet it PR, I'm implementing logic. We, of course can write our own parser, but I'm trying to avoid reinventing wheel here. Another reason I'm using go-chi is because it's 100% compatible with `http/net` and has no other dependencies except this package.

File: pkg/pillar/cmd/zedagent/parseconfig_test.go [hunk 3]
```diff
@@ -1254,3 +1264,73 @@ func TestParseSRIOV(t *testing.T) {
 		},
 	}))
 }
+
+func TestParsePatchEnvelope(t *testing.T) {
+	g := NewGomegaWithT(t)
+	getconfigCtx := initGetConfigCtx(g)
+
```
Comment (christoph-zededa) [inline pkg/pillar/cmd/zedagent/parseconfig_test.go:1271]:
```suggestion
t.Parallel()
```
I am wondering why yetus is not complaining ...

Is gomega doing this under the hood?

Comment (uncleDecart) [reply pkg/pillar/cmd/zedagent/parseconfig_test.go:1271]: AFAIK gomega is not testing framework, ginko is. So I'm not sure it's doing this under the hood...

File: pkg/pillar/cmd/zedrouter/metadata.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (eriknordmark) [reply pkg/pillar/cmd/zedrouter/metadata.go:13]: Makes sense.

File: pkg/pillar/types/patchenvelopestypes.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+// PatchEnvelopeInfo - information
+// about patch envelopes
+type PatchEnvelopeInfo struct {
+	PatchId     string       `json:"patch-id"`
+	BinaryBlobs []BinaryBlob `json:"binary-blobs"`
+}
+
+func GetZipArchive(root string, pe PatchEnvelopeInfo) (string, error) {
```
Comment (milan-zededa) [inline pkg/pillar/types/patchenvelopestypes.go:47]: GetZipArchive should go into `pillar/utils`, `types` package should preferably contain only global type declarations and their methods.

File: pkg/pillar/cmd/zedrouter/zedrouter.go [hunk 5]
```diff
@@ -776,6 +781,20 @@ func (z *zedrouter) initSubscriptions() (err error) {
 	if err != nil {
 		return err
 	}
+
+	// Information about patch envelopes
+	z.subPatchEnvelopeInfo, err = z.pubSub.NewSubscription(pubsub.SubscriptionOptions{
+		AgentName:   "zedagent",
+		MyAgentName: agentName,
+		TopicImpl:   types.PatchEnvelopes{},
+		Activate:    false,
```
Comment (milan-zededa) [inline pkg/pillar/cmd/zedrouter/zedrouter.go:790]: Isn't it necessary to activate the subscription eventually?

File: pkg/pillar/persistcache/README.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
+# Persisted Cache (persistcache)
+
+## Use-case
+
+This package implements a persistance cache. The main use-case is to store data on filesystem and
```
Comment (milan-zededa) [inline pkg/pillar/persistcache/README.md:5]: Would be good to mention that the cache has key-value semantics.

File: pkg/pillar/cmd/zedrouter/metadata.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (uncleDecart) [reply pkg/pillar/cmd/zedrouter/metadata.go:13]: So @christoph-zededa we do not use gorilla mux explicitly in EVE, however it's indirect dependency from gorilla ws, which is not EOL and also used to communicate with controller and in controller we are still using gorilla ws, so I suggest we keep using gorilla ws for now

File: pkg/pillar/cmd/zedagent/parseconfig_test.go [hunk 7]
```diff
@@ -1254,3 +1264,73 @@ func TestParseSRIOV(t *testing.T) {
...
+						ArtifactMetaData: &artiactMetadata,
+					},
+				},
+				AppInstIdsAllowed: []string{appU1, appU2},
+			},
+		},
+	}
+
+	path, _ := os.Getwd()
+	persistCacheFolder := filepath.Join(path, "testPersist/")
```
[the remark this answers is not in the document]
Comment (uncleDecart) [reply pkg/pillar/cmd/zedagent/parseconfig_test.go:1310]: Makes sense

File: pkg/pillar/cmd/zedagent/parsepatchenvelopes.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+	"fmt"
+
+	"crypto/sha256"
+
+	zconfig "github.com/lf-edge/eve-api/go/config"
+	"github.com/lf-edge/eve/pkg/pillar/persistcache"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+)
+
+const PERSIST_CACHE_FILEPATH = "/persist/cache/"
```
[the remark this answers is not in the document]
Comment (uncleDecart) [reply pkg/pillar/cmd/zedagent/parsepatchenvelopes.go:17]: Will change it

File: pkg/pillar/persistcache/README.md [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+as string. This was developed for configuration services feature in EVE. Since there is an option for 
+user to provide object to store via inline query, EVE has to store it even after reboot.
+
+## API & Usage
+
+Main structure is `pesistCache` outside of `persistcache` package it could be created via `Load` function:
+it is loads values from cache if they are present or created folder if there was none.
+`persistCache` structure has 3 operations:
+
+- *Get* value from in-memory cache
```
[the remark this answers is not in the document]
Comment (uncleDecart) [reply pkg/pillar/persistcache/README.md:16]: Will add this

File: pkg/pillar/persistcache/README.md (hunk 6 above)
Comment (uncleDecart) [reply pkg/pillar/persistcache/README.md:5]: Fair :)

File: pkg/pillar/types/patchenvelopestypes.go (hunk 4 above)
Comment (uncleDecart) [reply pkg/pillar/types/patchenvelopestypes.go:47]: We can change signature to :D 
```go
func (pe PatchEnvelopeInfo) GetZipArchive(root string) (string, error) {
//...
}
```

File: pkg/pillar/cmd/zedagent/parsepatchenvelopes.go [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+// cacheInlineBinaryArtifact stores inline artifact as file and
+// returns path to it to be served by HTTP server
+func cacheInlineBase64Artifact(artifact *zconfig.InlineOpaqueBase64Data, persistCacheFilepath string) (*types.BinaryBlob, error) {
+	pc, err := persistcache.New(persistCacheFilepath)
+	if err != nil {
+		return nil, err
+	}
+
+	metadata := artifact.GetBase64MetaData()
+	data := artifact.GetBase64Data()
```
[the remark this answers is not in the document]
Comment (uncleDecart) [reply pkg/pillar/cmd/zedagent/parsepatchenvelopes.go:108]: We have two possible entities as patchEnvelopes: inline objects and externally referenced objects. Inline objects we receive as part of `EdgeDevConfig` and they are limited to 64Kbytes.
If there's anything more than 64Kbytes it should be created as externally referenced object. That is simply Volume referenced in patchEnvelope

File: pkg/pillar/persistcache/persistcache.go [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+)
+
+type persistCache struct {
+	sync.Mutex
+	cache map[string]string
```
[the remark this answers is not in the document]
Comment (uncleDecart) [reply pkg/pillar/persistcache/persistcache.go:16]: Inline patch envelope we get from `EdgeDevConfig` should be stored somewhere and then served directly to client. We just need to keep track of all files created, deleted, modified. Cache in this case is not that important.

File: pkg/pillar/cmd/zedagent/parseconfig_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (uncleDecart) [reply pkg/pillar/cmd/zedagent/parseconfig_test.go]: Done

File: pkg/pillar/cmd/zedagent/parsepatchenvelopes.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (uncleDecart) [reply pkg/pillar/cmd/zedagent/parsepatchenvelopes.go]: Moved it to type constants, done

File: pkg/pillar/cmd/zedrouter/metadata_handlers.go [hunk 12]
```diff
@@ -0,0 +1,625 @@
...
+			b, err := json.Marshal(envelopes)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusUnprocessableEntity)
+				return
+			}
+			w.WriteHeader(http.StatusOK)
+			w.Write(b)
+			return
+		} else {
+			http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
```
[the remark this answers is not in the document]
Comment (uncleDecart) [reply pkg/pillar/cmd/zedrouter/metadata_handlers.go]: done

File: pkg/pillar/cmd/zedrouter/zedrouter.go (hunk 5 above)
Comment (uncleDecart) [reply pkg/pillar/cmd/zedrouter/zedrouter.go:791]: Done

File: pkg/pillar/persistcache/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (uncleDecart) [reply pkg/pillar/persistcache/README.md]: updated

File: pkg/pillar/persistcache/README.md (hunk 6 above)
Comment (uncleDecart) [reply pkg/pillar/persistcache/README.md:5]: Updated

**Comment by uncleDecart** [comment]
Pushed new commits addressing some of the conversations in this PR and also adding support for external patch envelope objects

File: pkg/pillar/cmd/zedagent/parsepatchenvelopes.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (eriknordmark) [reply pkg/pillar/cmd/zedagent/parsepatchenvelopes.go:97]: @uncleDecart I think the whole EdgeDevConfig is limited to about 64Kbytes so if we want to support that size for each inlined object then I think we will have an issue? Makes sense to check the size limit for the EdgeDevConfig (if any) in eden and the Zededa commercial controller.

[the remark this answers is not in the document]
Comment (eriknordmark) [reply pkg/pillar/cmd/zedagent/parsepatchenvelopes.go:97]: I wouldn't be a bad idea to have the Put function (or a variant of it) take a bytes.Buffer so we don't have to read lots of 64kbyte items into memory merely to write them to a file.

File: pkg/pillar/cmd/zedrouter/pubsubhandlers.go [hunk 13]
```diff
@@ -634,3 +634,27 @@ func (z *zedrouter) handleAppInstDelete(ctxArg interface{}, key string,
...
+	statusArg interface{}) {
+}
+
+func (z *zedrouter) handleVolumeStatusModify(ctxArg interface{}, key string,
+	statusArg interface{}, oldStatusArg interface{}) {
+}
+
+func (z *zedrouter) handleVolumeStatusDelete(ctxArg interface{}, key string,
+	statusArg interface{}) {
+}
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/pubsubhandlers.go:660]: Are there functions supposed to be empty? If so you don't need to register functions in NewSubscription().

File: pkg/pillar/persistcache/README.md [hunk 14]
```diff
@@ -0,0 +1,36 @@
+# Persisted Cache (persistcache)
+
+## Use-case
+
+This package implements a persistance cache. The main use-case is to store data on filesystem and
+provide access to it via API. Limitation of current implementation is that data should be represented
+as string. This was developed for configuration services feature in EVE. Since there is an option for
```
Comment (eriknordmark) [inline pkg/pillar/persistcache/README.md:7]: string implies valid UTF-8. Can't you define it to be a byte slice instead? That is much more general.

File: pkg/pillar/persistcache/README.md [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+provide access to it via API. Limitation of current implementation is that data should be represented
+as string. This was developed for configuration services feature in EVE. Since there is an option for 
+user to provide object to store via inline query, EVE has to store it even after reboot. 
+
+From library user perspective this library works as `map[string]string` with respective files created for each key in root folder.
+
+## API & Usage
+
+Main structure is `pesistCache` outside of `persistcache` package it could be created via `Load` function:
+it is loads values from cache if they are present or created folder if there was none.
```
Comment (eriknordmark) [inline pkg/pillar/persistcache/README.md:15]: Hard to parse. Is there a type about should say "create folder"?

File: pkg/pillar/persistcache/README.md [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+
+### Notes:
+- persistCache create lock file which does not allow using same folder in different threads, goroutines, apps, before creating instance on the same resource one should *Close* first instance. In case of multithreading use it is advised to use same object (basic mutex is implemented, so persistCache is not lock-free object, but you stil can share it between goroutines)
+- Lazy initalisation is used. That means no values are loaded during *New* call, rather *Get* will load value from filesystem if it is not in the cache
+
+
+## Design decisions
+
+### Why we are storing separate files and not saving whole structure as file?
++ If objects stored are large it takes less time to save/update them and less code
```
Comment (eriknordmark) [inline pkg/pillar/persistcache/README.md:31]: Presumably it also avoids a Put of one key from affecting the storage of another key, which could happen if it is a single file and the device is powered off before everything has been sync'ed to disk.

File: pkg/pillar/persistcache/persistcache.go [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+
+	return filepath, nil
+}
+
+// Remove element from cache and filesystem
+func (pc *persistCache) Delete(key string) error {
+	pc.Lock()
+	defer pc.Unlock()
+
+	if _, err := os.Stat(filepath.Join(pc.tempDir, LOCK_FILE_NAME)); os.IsNotExist(err) {
```
Comment (eriknordmark) [inline pkg/pillar/persistcache/persistcache.go:162]: Why do we need a lock file? Raises complexity  around cleaning up after a reboot and there is still a race between a Stat and a Create call.

With WriteRename as the underlying mechanism for modifying the files then even if multiple users (processes, goroutines) use persistcache, a Get of a key will either fetch the content for that key before a Put or after but not mixed/partial content. That should be sufficient AFAIU.

File: pkg/pillar/persistcache/persistcache.go [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+		return &LockFileNotFound{}
+	}
+
+	delete(pc.cache, key)
+	return os.Remove(filepath.Join(pc.root, key))
+}
+
+func isValidKey(key string) bool {
+	if strings.Contains(key, "/") {
```
Comment (eriknordmark) [inline pkg/pillar/persistcache/persistcache.go:171]: What about names like ".." and "."?
I think it makes sense to call path.Clean() on the key (since it will replace .. with ../ etc).
Then you can check for "/".

File: pkg/pillar/persistcache/persistcache_test.go [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+
+package persistcache_test
+
+import (
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/lf-edge/eve/pkg/pillar/persistcache"
+	"gotest.tools/assert"
```
Comment (eriknordmark) [inline pkg/pillar/persistcache/persistcache_test.go:12]: Is there something missing in the test package we use in all the other unit tests?

**Comment by uncleDecart** [comment]
Renaming it back to inline objects, will add external objects and statuses in different PRs. Updated commits to include only code for inline binary artifacts, added documentation for Patch Envelopes and updated code from PR suggestions. I think we can merge this PR if there are no more comments or remarks

File: pkg/pillar/persistcache/persistcache_test.go (hunk 19 above)
Comment (uncleDecart) [reply pkg/pillar/persistcache/persistcache_test.go]: No, I removed assert completely and used gomega

File: pkg/pillar/persistcache/persistcache.go (hunk 18 above)
Comment (uncleDecart) [reply pkg/pillar/persistcache/persistcache.go:126]: Makes sense, added it

File: pkg/pillar/persistcache/persistcache.go (hunk 17 above)
Comment (uncleDecart) [reply pkg/pillar/persistcache/persistcache.go]: Yes, you're right, let's keep it simple since WriteRename will be enough

File: pkg/pillar/persistcache/README.md (hunk 14 above)
Comment (uncleDecart) [reply pkg/pillar/persistcache/README.md]: Makes sense I changed it to byte slice

File: pkg/pillar/cmd/zedrouter/pubsubhandlers.go (hunk 13 above)
Comment (uncleDecart) [reply pkg/pillar/cmd/zedrouter/pubsubhandlers.go]: Those functions are needed for external patch envelopes, removed declarations from this PR, will introduce them in separate PR

File: pkg/pillar/persistcache/persistcache.go [hunk 20]
```diff
@@ -0,0 +1,188 @@
+package persistcache
```
[the remark this answers is not in the document]
Comment (uncleDecart) [reply pkg/pillar/persistcache/persistcache.go:1]: Added it

File: pkg/pillar/cmd/zedagent/parsepatchenvelopes.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (uncleDecart) [reply pkg/pillar/cmd/zedagent/parsepatchenvelopes.go:97]: @eriknordmark, checked size on eden and Zededa commercial controller, we don't have any 64Kbytes restriction on EdgeDevConfig

**Comment by uncleDecart** [comment]
Updated documentation, fixed some of the yetus warnings

File: pkg/pillar/persistcache/README.md (hunk 15 above)
Comment (uncleDecart) [reply pkg/pillar/persistcache/README.md]: Rephrased it

File: pkg/pillar/persistcache/README.md (hunk 16 above)
Comment (uncleDecart) [reply pkg/pillar/persistcache/README.md:47]: added it

File: pkg/pillar/cmd/zedagent/parsepatchenvelopes.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (eriknordmark) [reply pkg/pillar/cmd/zedagent/parsepatchenvelopes.go:97]: Thanks

File: pkg/pillar/cmd/zedrouter/metadata_handlers.go [hunk 21]
```diff
@@ -0,0 +1,693 @@
...
+	}
+}
+
+func sendError(w http.ResponseWriter, code int, msg string) {
+	w.Header().Set("Content-Type", "text/plain")
+	w.WriteHeader(code)
+	w.Write([]byte(fmt.Sprintf("{\"message\": \"%s\"}", msg)))
+}
+
+func HandlePatchDownload(z *zedrouter) func(http.ResponseWriter, *http.Request) {
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/metadata_handlers.go:596]: Still some revive and codespell issues from yetus to fix.

File: pkg/pillar/types/patchenvelopestypes_test.go [hunk 22]
```diff
@@ -0,0 +1,113 @@
+package types
```
Comment (eriknordmark) [inline pkg/pillar/types/patchenvelopestypes_test.go:1]: Also missing Copyright and License lines.

File: pkg/pillar/cmd/zedrouter/pubsubhandlers.go (hunk 13 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/zedrouter/pubsubhandlers.go]: ok

File: pkg/pillar/persistcache/persistcache.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (eriknordmark) [reply pkg/pillar/persistcache/persistcache.go:1]: Did you push that change? Doesn't show in github.

File: pkg/pillar/persistcache/persistcache.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (eriknordmark) [reply pkg/pillar/persistcache/persistcache.go]: A cache makes sense, but we don't need to read data into memory if you make Get() return an io.Reader (and make Put take an io.Reader). Then you can do an io.Copy  which will use a limited size buffer.

File: pkg/pillar/persistcache/README.md [hunk 23]
```diff
@@ -0,0 +1,31 @@
+# Persisted Cache (persistcache)
+
+## Use-case
+
+This package implements a persistance cache. The main use-case is to store data on filesystem and
+provide access to it via API. Limitation of current implementation is that data should be represented
+as string. This was developed for configuration services feature in EVE. Since there is an option for 
+user to provide object to store via inline query, EVE has to store it even after reboot.
```
[the remark this answers is not in the document]
Comment (uncleDecart) [reply pkg/pillar/persistcache/README.md]: Added doc for it

File: pkg/pillar/persistcache/persistcache.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (uncleDecart) [reply pkg/pillar/persistcache/persistcache.go:4]: Now you should see it :)

File: pkg/pillar/cmd/zedrouter/metadata_handlers.go (hunk 21 above)
Comment (uncleDecart) [reply pkg/pillar/cmd/zedrouter/metadata_handlers.go:602]: Updated

File: pkg/pillar/types/patchenvelopestypes_test.go (hunk 22 above)
Comment (uncleDecart) [reply pkg/pillar/types/patchenvelopestypes_test.go:4]: Now you can see it :)

**Comment by uncleDecart** [comment]
Struggle goes on, fixed yetus issues and updates license files.

In addition, now inline objects are cleared every time EdgeDevConfig gets created/recreated 

Have yet to implement io.Reader for persistcache

File: pkg/pillar/persistcache/persistcache.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (uncleDecart) [reply pkg/pillar/persistcache/persistcache.go]: if we don't use WriteRename in Put, because there we will have to convert to slice of bytes which will load whole object in memory

[the remark this answers is not in the document]
Comment (uncleDecart) [reply pkg/pillar/persistcache/persistcache.go]: @eriknordmark using Reader also implies that you read only once. And passing by value does not copy whole stream, I.e. 

```
type MyFavStruct struct {
  Field map[string]io.Reader
}

func main() {

s := strings.NewReader( "bazinga" )

a := MyFavStruct{
  Field: make(map[string]io.Reader)
}

a.Field["bobo"] = s
sr1 := new(strings.Builder)

io.Copy(sr1, s)
fmt.PrintIn(sr1.String())

sr2 := new(strings.Builder)
io.Copy(sr2, a.Field["bobo"])


fmt.Printin(sr2.String())

}

```

Will output bazinga and empty line 

I think we should stay with []byte or wrap it around bytes buffer for chunking.

Current implementation with cache and its usage won't load more than one object to byte slice.

**Review by eriknordmark — COMMENTED** [review]
> This, of course create a possible state where app instance tries
to access file which is not there anymore and new config has not
been propagated yet to zedrouter, but in that case metadata server
will return error and app instance can either retry or request
description.json once again

What happens to the description.json we serve to the app instance as we do these delete and recreate? Could it still list the files but a fetch of those files might fail? Or will it exclude the files which are deleted until they have been recreated?
I don't know how likely app developers and other tools are to trip on this.
Couldn't we have the running EVE keep the sha256 of each inline object in memory and compare the patch envelopes against that, and only do the delete+recreate if a particular inline object has changed?

File: pkg/pillar/types/patchenvelopestypes_test.go (hunk 22 above)
Comment (eriknordmark) [reply pkg/pillar/types/patchenvelopestypes_test.go:4]: yes.

File: pkg/pillar/persistcache/persistcache.go [hunk 24]
```diff
@@ -0,0 +1,146 @@
...
+	fileutils "github.com/lf-edge/eve/pkg/pillar/utils/file"
+)
+
+type persistCache struct {
+	sync.Mutex
+	cache map[string][]byte
+	root  string
+}
+
+const FileMask = 0755
```
Comment (eriknordmark) [inline pkg/pillar/persistcache/persistcache.go:23]: Do you need this to be exported?

File: pkg/pillar/persistcache/persistcache.go [hunk 25]
```diff
@@ -0,0 +1,146 @@
...
+const LockFileName = "persistcache.lock"
```
Comment (eriknordmark) [inline pkg/pillar/persistcache/persistcache.go:24]: I thought you removed the lock file, but it is still here and in the test

File: pkg/pillar/cmd/zedrouter/zedrouter.go [hunk 26]
```diff
@@ -776,6 +796,33 @@ func (z *zedrouter) initSubscriptions() (err error) {
...
+		WarningTime: warningTime,
+		ErrorTime:   errorTime,
+	})
+	if err != nil {
+		return err
+	}
+
+	// Infromation about volumes reffered in external patch envelopes
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/zedrouter.go:813]: More yetus

File: pkg/pillar/cmd/zedrouter/metadata_handlers.go (hunk 21 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/zedrouter/metadata_handlers.go:602]: But still more below.

**Comment by uncleDecart** [comment]
Fighting with yetus yet again :) 

Addressed comments added

File: pkg/pillar/persistcache/persistcache.go (hunk 24 above)
Comment (uncleDecart) [reply pkg/pillar/persistcache/persistcache.go]: No, I changed it

File: pkg/pillar/persistcache/persistcache.go (hunk 25 above)
Comment (uncleDecart) [reply pkg/pillar/persistcache/persistcache.go]: removed it all

File: pkg/pillar/persistcache/persistcache.go [hunk 27]
```diff
@@ -0,0 +1,146 @@
...
+
+type InvalidKeyError struct{}
```
[the remark this answers is not in the document]
Comment (uncleDecart) [reply pkg/pillar/persistcache/persistcache.go:34]: fixed

File: pkg/pillar/cmd/zedrouter/zedrouter.go (hunk 26 above)
Comment (uncleDecart) [reply pkg/pillar/cmd/zedrouter/zedrouter.go]: fixed

File: pkg/pillar/cmd/zedrouter/metadata_handlers.go (hunk 21 above)
Comment (uncleDecart) [reply pkg/pillar/cmd/zedrouter/metadata_handlers.go:602]: :smiling_face_with_tear:

Comment (uncleDecart) [reply pkg/pillar/cmd/zedrouter/metadata_handlers.go:602]: Let's see what yetus have to say now

File: pkg/pillar/cmd/zedrouter/metadata_handlers.go [hunk 28]
```diff
@@ -0,0 +1,713 @@
...
+			peObj, err := z.subPatchEnvelopeInfo.Get("zedagent")
+			if err != nil {
+				sendError(w, http.StatusNoContent, "Cannot get patch envelope subscription")
+				return
+			}
+
+			pe := peObj.(types.PatchEnvelopes)
+
+			remoteIP := net.ParseIP(strings.Split(r.RemoteAddr, ":")[0])
+			anStatus := z.lookupAppNetworkStatusByAppIP(remoteIP)
```
Comment (gkodali-zededa) [inline pkg/pillar/cmd/zedrouter/metadata_handlers.go:692]: Do you check if the given app instance (caller of API) is part of the allowed list before reaching here?

File: pkg/pillar/cmd/zedagent/parsepatchenvelopes.go [hunk 29]
```diff
@@ -0,0 +1,153 @@
...
+	if err != nil {
+		log.Errorf("Failed to load persistCache %v", err)
+		return
+	}
+	blobsBefore := pc.Objects()
+
+	var blobsAfter []string
+	patchEnvelopes := config.GetPatchEnvelopes()
+	result := types.PatchEnvelopes{}
+	for _, pe := range patchEnvelopes {
```
Comment (gkodali-zededa) [inline pkg/pillar/cmd/zedagent/parsepatchenvelopes.go:46]: Do we destroy and re-create patch envelopes every time configuration gets pulled from zedcloud? Does it happen even when the patch envelope did not change?

File: pkg/pillar/cmd/zedrouter/metadata_handlers.go (hunk 28 above)
Comment (uncleDecart) [reply pkg/pillar/cmd/zedrouter/metadata_handlers.go:692]: I follow same logic as other handlers in resolving remoteAddr to uuid, then check if there are any patch envelopes available for given app instance uuid and return them

File: pkg/pillar/cmd/zedagent/parsepatchenvelopes.go (hunk 29 above)
Comment (uncleDecart) [reply pkg/pillar/cmd/zedagent/parsepatchenvelopes.go:46]: We destroy only inline objects, which are not referenced anymore when new configuration comes. If inline patch envelope did not change we will not override it.

File: docs/ECO-METADATA.md [hunk 30]
```diff
@@ -311,3 +311,54 @@ EVE is generating diagnostic output on a console (if there is one) which summari
...
+Get list of available Patch Envelopes `/eve/v1/patch/description.json`
+For example:
+
+```bash
+curl -X GET -v http://169.254.169.254/eve/v1/patch/description.json
+[
+
+    {
+        "AllowedApps":[
+            "17daa0ff-39d6-42be-a537-44c974276aec"
```
[the remark this answers is not in the document]
Comment (uncleDecart) [reply docs/ECO-METADATA.md]: Updated

**Comment by uncleDecart** [comment]
Addressed review comments

File: docs/ECO-METADATA.md [hunk 31]
```diff
@@ -311,3 +311,51 @@ EVE is generating diagnostic output on a console (if there is one) which summari
...
+```bash
+curl -X GET -v http://169.254.169.254/eve/v1/patch/description.json
+[
+
+    {
+        "PatchId":"699fbdb2-e455-448f-84f5-68e547ec1305",
+        "BinaryBlobs":[
+            {
+                "file-name":"textfile1.txt",
+                "file-sha":"e5f3f80da0f9a3add3540a80d214069079f71e879b87c51c4c3a258989294812",
```
Comment (eriknordmark) [inline docs/ECO-METADATA.md:333]: Is there some comment to make detsecrets shut up for this line?

**Comment by uncleDecart** [comment]
@eriknordmark I think that this PR is ready to be merged (there are some yetus errors still, but I don't think we should address them) let me know what you think

**Review by eriknordmark — APPROVED** [review]
LGTM but still plenty of yetus issues to fix.

File: pkg/pillar/types/patchenvelopestypes.go (hunk 4 above)
Comment (eriknordmark) [reply pkg/pillar/types/patchenvelopestypes.go:87]: FWIW changing the signature makes sense; in any case this operators on a PE.

File: pkg/pillar/persistcache/persistcache.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (eriknordmark) [reply pkg/pillar/persistcache/persistcache.go]: Ok

File: pkg/pillar/types/patchenvelopestypes.go (hunk 4 above)
Comment (milan-zededa) [reply pkg/pillar/types/patchenvelopestypes.go:87]: @uncleDecart It is not about the signature. Package `types` should only contain structures defining APIs between pillar microservices. This is imported by every microservice, including "mmagent" in the future. Having this method here will add dependencies and increase binary size (which is unnecessary for everything outside of zedbox, i.e. edgeview, mmagent, etc.).

Comment (uncleDecart) [reply pkg/pillar/types/patchenvelopestypes.go:87]: Moved this code to utils

**Comment by uncleDecart** [comment]
@eriknordmark all checks are green, I added ignoring of high entropy strings for tests and removed sha from documentation examples, rest I fixed
